core/services/sleeper_task_test.go

```diff
@@ -1,187 +1,93 @@
 	"github.com/stretchr/testify/require"
 )
 
-type testWorker struct {
-	output chan struct{}
+type countingWorker struct {
+	numJobsPerformed int32
+	delay            time.Duration
 }
 
-func (t *testWorker) Work() {
-	t.output <- struct{}{}
+func (t *countingWorker) Work() {
+	if t.delay != 0 {
+		time.Sleep(t.delay)
+	}
+	// Without an atomic, the race detector fails
+	atomic.AddInt32(&t.numJobsPerformed, 1)
 }
 
-type longRunningWorker struct {
-	output   chan struct{}
-	finished bool
+func (t *countingWorker) getNumJobsPerformed() int {
+	return int(atomic.LoadInt32(&t.numJobsPerformed))
 }
 
-func (w *longRunningWorker) Work() {
-	w.output <- struct{}{}
-	time.Sleep(100 * time.Millisecond)
-	w.finished = true
-}
+func TestSleeperTask_WakeupAfterStopPanics(t *testing.T) {
+	t.Parallel()
 
-func TestSleeperTask(t *testing.T) {
-	worker := testWorker{output: make(chan struct{})}
-	sleeper := services.NewSleeperTask(&worker)
-
-	sleeper.Start()
-	sleeper.WakeUp()
-
-	gomega.NewGomegaWithT(t).Eventually(worker.output).Should(gomega.Receive(&struct{}{}))
+	worker := &countingWorker{}
+	sleeper := services.NewSleeperTask(worker)
 
 	sleeper.Stop()
-}
-
-func TestSleeperTask_WakeupBeforeStarted(t *testing.T) {
-	worker := testWorker{output: make(chan struct{})}
-	sleeper := services.NewSleeperTask(&worker)
 
-	sleeper.WakeUp()
-	sleeper.Start()
-
-	gomega.NewGomegaWithT(t).Eventually(worker.output).Should(gomega.Receive(&struct{}{}))
-
-	sleeper.Stop()
+	require.Panics(t, func() {
+		sleeper.WakeUp()
+	})
+	gomega.NewGomegaWithT(t).Eventually(worker.getNumJobsPerformed).Should(gomega.Equal(0))
 }
 
-func TestSleeperTask_WakeupAfterStarted(t *testing.T) {
-	worker := testWorker{output: make(chan struct{})}
-	sleeper := services.NewSleeperTask(&worker)
-
-	sleeper.Start()
-	sleeper.WakeUp()
-
-	gomega.NewGomegaWithT(t).Eventually(worker.output).Should(gomega.Receive(&struct{}{}))
+func TestSleeperTask_CallingStopTwicePanics(t *testing.T) {
+	t.Parallel()
 
+	worker := &countingWorker{}
+	sleeper := services.NewSleeperTask(worker)
 	sleeper.Stop()
+	require.Panics(t, func() {
+		sleeper.Stop()
+	})
 }
 
-func TestSleeperTask_WakeupAfterStopped(t *testing.T) {
-	worker := testWorker{output: make(chan struct{})}
-	sleeper := services.NewSleeperTask(&worker)
-
-	sleeper.Start()
-	sleeper.Stop()
+func TestSleeperTask_WakeupPerformsWork(t *testing.T) {
+	t.Parallel()
 
-	err := sleeper.WakeUp()
+	worker := &countingWorker{}
+	sleeper := services.NewSleeperTask(worker)
 
-	assert.Error(t, err)
+	sleeper.WakeUp()
+	gomega.NewGomegaWithT(t).Eventually(worker.getNumJobsPerformed).Should(gomega.Equal(1))
+	sleeper.Stop()
 }
 
-func TestSleeperTask_WakeupNotBlockedWhileWorking(t *testing.T) {
-	worker := testWorker{output: make(chan struct{})}
-	sleeper := services.NewSleeperTask(&worker)
+func TestSleeperTask_WakeupEnqueuesExactlyTwiceWithoutBlocking(t *testing.T) {
+	t.Parallel()
 
-	sleeper.Start()
+	worker := &countingWorker{delay: 200 * time.Millisecond}
```

Comment:
Is there a way to do it with locks or channels instead of a timed delay?

Reply:
Parallel tests can sometimes take a pathologically long time (especially on a slow machine like mine) so this sort of thing makes me wary because it can cause flakiness and indeterministic tests which is something I very much want to move us away from.